cloudinit/sources/__init__.py

```diff
@@ -307,6 +307,29 @@ def _unpickle(self, ci_pkl_version: int) -> None:
     def __str__(self):
         return type_utils.obj_name(self)
 
+    def ds_detect(self) -> bool:
+        """Check if running on this datasource"""
+        return True
+
+    def override_ds_detect(self):
+        """Override if either:
+        - only a single datasource defined (nothing to fall back to)
+        - TODO: commandline argument is used (ci.ds=OpenStack)
```

Comment:
Either `ci.ds=OpenStack` or `ci.datasource=OpenStack` result in `sys_cfg["datasource_list] == ["OpenStack", "None"]`

Reply:
For future, not this PR. I don't like that NoCloud datasource is unique in that it [accepts `ds=nocloud-*` params](https://github.com/canonical/cloud-init/blob/main/tools/ds-identify#L838-L841) but every other datasource has to rely on ci.ds or ci.datasource. That only adds to confusion.

Reply:
> but maybe cloud-init it's a bit presumptive to assume we can own the ds= scope of kernel parameter

Agreed. That, and the fact that `ci.ds=` and `ci.datasource=` have parity, I don't see any reason not to deprecate `ci.ds=` (and only documenting `ci.datasource=`) when we get around to making this changes.

Reply:
to clarify, I mean I think it'd be ideal to support `ds=openstack`  because that means typing less on the grub menu  when manually extending supplemental kernel params.... so I'd want to avoid dropping `ci.ds=` because less character is better where possible. I was just uncertain if we'd want to support even shorter alias `ds=` which would be 'outside` the top-level `ci` (cloud-init) scope.

Reply:
Eh, not sure I'm on board with this angle.

I lean towards the safety of longer & namespaced over shaving a few keystrokes, and less supported argument formats rather than more. More means more current and future potential collisions, more things to maintain, etc.

As we've seen with Ironic, different datasources make different usage of the kernel commandline, so we might not know of collisions on other datasources until we enable it on those datasources.

I'm willing to be convinced, my gut reaction is just the opposite of the proposed.

Reply:
> For future, not this PR. I don't like that NoCloud datasource is unique in that it [accepts ds=nocloud-* params](https://github.com/canonical/cloud-init/blob/main/tools/ds-identify#L838-L841) but every other datasource has to rely on ci.ds or ci.datasource. That only adds to confusion.

As for this earlier comment, I 100% agree. I'd just love to see NoCloud change to line up with the rest of cloud-init rather than move the rest of the datasources to line up with it.

Reply:
> I lean towards the safety of longer & namespaced over shaving a few keystrokes

I would normally agree, but I'm pretty sure there can still be fairly low character limits for the kernel command line. I think it's one of the few places that terseness should be the norm. I agree that collisions might be an issue though.

Reply:
> I would normally agree, but I'm pretty sure there can still be fairly low character limits for the kernel command line. I think it's one of the few places that terseness should be the norm. I agree that collisions might be an issue though.

Fair point. [On x86 the buffer is 2048 bytes](https://elixir.bootlin.com/linux/v6.2.1/source/arch/x86/include/asm/setup.h#L7), and poking around it looks like any architecture we should care about has a size of >=1024 bytes.

@blackboxsw @TheRealFalcon This is going to be documented in a followup PR to document datasource selection behavior. Do we have strong feelings about shaving those first three bytes (`ci.`)? Or does using `ci.ds=` as a tradeoff seem reasonable - shorter, but keep the name-spacing for more uniqueness?

Reply:
`ci.ds=` sounds good to me

Reply:
@TheRealFalcon @blackboxsw While documenting this change, I notice that we recently deprecated `url` in favor of `cloud-config-url` in 8e26b2f8a9ce8c1186a6e5f2e9bd90a984fde52a. For consistency, I'm thinking we should maybe revert this, and possibly switch to `ci.url=` while deprecating the other two.